src/object/tcb.c

```diff
@@ -1251,17 +1251,31 @@ exception_t decodeSetMCPriority(cap_t cap, word_t length, word_t *buffer)
 }
 
 #ifdef CONFIG_KERNEL_MCS
-exception_t decodeSetTimeoutEndpoint(cap_t cap, cte_t *slot)
+exception_t decodeSetTimeoutEndpoint(cap_t cap, word_t length, cte_t *slot, word_t *buffer)
 {
-    if (current_extra_caps.excaprefs[0] == NULL) {
-        userError("TCB SetSchedParams: Truncated message.");
+    if (length < 2 || current_extra_caps.excaprefs[0] == NULL) {
+        userError("TCB SetTimeoutEndpoint: Truncated message.");
+        current_syscall_error.type = seL4_TruncatedMessage;
         return EXCEPTION_SYSCALL_ERROR;
     }
 
+    word_t thData = getSyscallArg(0, buffer);
+    word_t thRights = getSyscallArg(1, buffer);
     cte_t *thSlot = current_extra_caps.excaprefs[0];
     cap_t thCap   = current_extra_caps.excaprefs[0]->cap;
 
     /* timeout handler */
+    if (thData != 0 || thRights != 0) {
+        thCap = updateCapData(false, thData, thCap);
+        thCap = maskCapRights(rightsFromWord(thRights), thCap);
+    }
+
+    deriveCap_ret_t dc_ret = deriveCap(thSlot, thCap);
+    if (dc_ret.status != EXCEPTION_NONE) {
+        return dc_ret.status;
+    }
+    thCap = dc_ret.cap;
+
```

Comment:
Maybe add a helper function for this block. Except if I missed something, the other two blocks do exactly the same.

Reply:
Good idea, thank you! That should also make verification easier in the future.

It wouldn't be part of this PR, but for the same reasons I will probably also want to add a helper function for the block that decodes the cspace and vspace parameters. I think that block is the same in `decodeSetSpace` and `decodeTCBConfigure`.